### PR TITLE
Add optional --stats flag to compile command

### PR DESCRIPTION
**Pull request by samutamm**
Compile flag uploads the given appid and source, compiles it and prints the sample to terminal.

Add new flag `--stats` that parses the intents and entities of the same sample and prints the distributions to the terminal.

For example, when trying this out inside a folder with a yaml file containing [the home automation example](https://docs.speechly.com/slu-examples/example-configuration/#home-automation) and having already created an app with appid `01b...c07`,
the command, `speechly compile -a 01b...c07 . --stats` gives following output:

```
INTENT		COUNT	PROPORTION
turn_off	24	0.521739
turn_on		22	0.478261

ENTITY TYPE	COUNT	PROPORTION
location	81	0.547297
device		67	0.452703

ENTITY VALUE	COUNT	PROPORTION
radio		17	0.114865
living room	12	0.081081
every device	12	0.081081
lights		12	0.081081
everything	12	0.081081
closet		10	0.067568
everywhere	10	0.067568
every room	10	0.067568
all devices	9	0.060811
all rooms	8	0.054054
bedroom		8	0.054054
kitchen		8	0.054054
pantry		8	0.054054
hallway		7	0.047297
television	5	0.033784
```
here the batch size was 32 examples.

### CODE REVIEW

**Comment by aukkonen** [comment]
How are the proportions calculated? I think in the example you chose, each utterance always contains some of the entities as well. How do the numbers look like if there are utterances that have no entities at all? I think it's okay to normalise them so that we always get the proportion out of ALL GENERATED utterances. And that's probably what you did here too, just checking to make sure.

Another question: Given that we're not supposed to sample entity values more or less uniformly, it's a bit strange that television and radio have such different probabilities here. Or is this because of the list of entity values being too short to qualify for the round-robin sampling algorithm?

**Comment by samutamm** [comment]
> How are the proportions calculated? I think in the example you chose, each utterance always contains some of the entities as well. How do the numbers look like if there are utterances that have no entities at all? I think it's okay to normalise them so that we always get the proportion out of ALL GENERATED utterances. And that's probably what you did here too, just checking to make sure.
> 

I normalised the intents by total number of intents (there were 46 in 32 utterances) and entity types and values by total number of entities (148 in 32 utterances). The counts include possible duplicates in one utterance. Is this a good way or should the entity counts tell more like 'how many of the utterances had this entity'? Then the entity counts should be divided by the number of utterances (here 32) and no counting the duplicate entities in one utterance.

> Another question: Given that we're not supposed to sample entity values more or less uniformly, it's a bit strange that television and radio have such different probabilities here. Or is this because of the list of entity values being too short to qualify for the round-robin sampling algorithm?

The sample did use round-robin but maybe it was too small to iterate all lists uniformly. To be seen whether same happens with n=1000. ([requires this](https://github.com/speechly/api/pull/9))

**Comment by bigdatabaracus** [comment]
This is great, though what i’d like to see here in terms of the distribution would be also an option to form thinks like this 
```
INTENT/ENTITY						COUNT	PROPORTION
turn_off(location=living room,device=radio)	12	0.221739
turn_on(location=bedroom,device=lights)		10	0.208261
.
.
.
```

**Comment by samutamm** [comment]
Here is an example of output with the same sal as previously

```
INTENTS		COUNT	DISTRIBUTION	AVG PER UTTRANCE
turn_on		761	0.509371	0.509371
turn_off	733	0.490629	0.490629

ENTITY TYPES	COUNT	DISTRIBUTION	AVG PER UTTRANCE
location	2320	0.523939	1.552878
device		2108	0.476061	1.410977

ENTITY VALUES	COUNT	DISTRIBUTION	AVG PER UTTRANCE
lights		375	0.084688	0.251004
every device	362	0.081752	0.242303
radio		360	0.081301	0.240964
...
living room	237	0.053523	0.158635
every room	221	0.049910	0.147925

ENTITY TYPES PER INTENT	COUNT	DISTRIBUTION	AVG PER UTTRANCE
turn_on(location=*)	1239	0.279810	0.829317
turn_off(location=*)	1081	0.244128	0.723561
turn_off(device=*)	1058	0.238934	0.708166
turn_on(device=*)	1050	0.237127	0.702811

ENTITY VALUES PER INTENT	COUNT	DISTRIBUTION	AVG PER UTTRANCE
turn_off(device=lights)		188	0.042457	0.125837
turn_on(device=lights)		187	0.042231	0.125167
turn_on(device=every device)	187	0.042231	0.125167
turn_off(device=radio)		186	0.042005	0.124498
turn_off(device=all devices)	181	0.040876	0.121151
...
turn_off(location=all rooms)	113	0.025519	0.075636
turn_on(location=every room)	107	0.024164	0.071620
turn_off(location=living room)	104	0.023487	0.069612

ENTITY VALUE PAIRS PER INTENT				COUNT	DISTRIBUTION	AVG PER UTTRANCE
turn_on(location=kitchen,device=all devices)		41	0.013469	0.027443
turn_on(location=kitchen,device=radio)			40	0.013141	0.026774
turn_off(location=every room,device=radio)		39	0.012812	0.026104
turn_on(location=closet,device=every device)		38	0.012484	0.025435
turn_on(location=every room,device=lights)		37	0.012155	0.024766
turn_off(location=pantry,device=lights)			37	0.012155	0.024766
turn_on(location=kitchen,device=lights)			37	0.012155	0.024766
turn_on(location=all rooms,device=every device)		36	0.011827	0.024096
turn_on(location=everywhere,device=every device)	36	0.011827	0.024096
turn_on(location=kitchen,device=television)		36	0.011827	0.024096
turn_on(location=hallway,device=everything)		35	0.011498	0.023427
turn_off(location=bedroom,device=all devices)		35	0.011498	0.023427
turn_on(location=everywhere,device=radio)		35	0.011498	0.023427
turn_off(location=everywhere,device=lights)		34	0.011170	0.022758
turn_on(location=pantry,device=lights)			34	0.011170	0.022758
...
turn_on(location=hallway,device=all devices)		20	0.006570	0.013387
turn_off(location=living room,device=everything)	20	0.006570	0.013387
turn_off(location=everywhere,device=everything)		19	0.006242	0.012718
turn_on(location=every room,device=radio)		17	0.005585	0.011379
turn_on(location=every room,device=all devices)		17	0.005585	0.011379
```

**Comment by aukkonen** [comment]
Maybe someone who's more knowledgeable in Go than me should take a quick look at this, but looks okay!

**Review by aukkonen — DISMISSED** [review]

File: cmd/common.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ type ValidateWriter struct {
 type CompileWriter struct {
 	appId  string
 	stream salv1.Compiler_CompileClient
+	batchSize int32
```
Comment (langma) [inline cmd/common.go:18]: This is not properly formatted, please run the sources through `gofmt` and maybe add that step to `Makefile` also.

**Review by langma — CHANGES_REQUESTED** [review]

**Review by langma — APPROVED** [review]

**Pull request merged**